File: agents/apps/ryu/am_interface.py
import socket
from os.path import expanduser
import sys
import struct

class AMInterface:
    def __init__(self, obj, logger):
        self.appAgent = obj
        self.logger = logger

    def setServerAddr(self):
        server_address = ('10.0.2.2', 3366) # default ip, port pair for vm
        home = expanduser("~")
        f = open(home + "/agent.cfg", "r")

        while True:
            line = f.readline()
            if not line: break

            if "MANAGER_IP" in line:
               manager_ip = line.split("=")[1].rstrip()
            elif "MANAGER_PORT" in line:
               manager_port = int(line.split("=")[1].rstrip())

        if manager_ip is not None and manager_port is not None:
            server_address = (manager_ip, manager_port)

        return server_address


    def connectServer(self, server_address):

        # Create a TCP/IP socket
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

        # Connect the socket to the port where the server is listening

        if not server_address:
            server_address = ('10.0.2.2', 3366)

        print '[AMInterface] Connecting to %s port %s' % server_address
        sock.connect(server_address)

        try:
            # Send AppAgent Message
            message = 'AppAgent'
            m = '\x00' + '\x08' + message
            self.logger.info("[AMInterface] Handshaking with AgentManager...")
            self.writeUTF(sock, message)

            # Receive OK Message
            data = self.readUTF(sock)
            self.logger.info("[AMInterface] Received from AgentManager: " + data)

            # Receive Code
            while True:
                data = self.readUTF(sock)
                self.logger.info("AgentManager: " + data)
                if "3.1.020" in data:
                    result = self.appAgent.testControlMessageDrop()
                    self.writeUTF(sock, result)
                elif "3.1.030" in data:
                    self.appAgent.testInfiniteLoops()
                elif "3.1.040" in data:
                    result = self.appAgent.testInternalStorageAbuse()
                    self.writeUTF(sock, result)
                elif "3.1.070" in data:
                    result = self.appAgent.testFlowRuleModification()
                    self.writeUTF(sock, result)
                elif "3.1.080" in data:
                    self.appAgent.testFlowTableClearance()
                elif "3.1.80" in data:
                    if "false" in data:
                        self.appAgent.callFlowTableClearance()
                elif "3.1.090" in data:
                    result = self.appAgent.testEventListenerUnsubscription()
                    self.writeUTF(sock, result)
                elif "3.1.100" in data:
                    result = self.appAgent.testApplicationEviction()
                    self.writeUTF(sock, result)
                elif "3.1.110" in data:
                    self.appAgent.testMemoryExhaustion()
                elif "3.1.120" in data:
                    self.appAgent.testCPUExhaustion()
                elif "3.1.130" in data:
                    self.appAgent.testSystemVariableManipulation()
                elif "3.1.140" in data:
                    self.appAgent.testSystemCommandExecution()
                elif "3.1.190" in data:
                    self.appAgent.testFlowRuleFlooding()
                elif "3.1.200" in data:
                    result = self.appAgent.testSwitchFirmwareMisuse()
                    self.writeUTF(sock, result)
                elif "2.1.060" in data:
                    result = self.appAgent.testUnFlaggedFlowRemoveMsgNotification()
                    self.writeUTF(result)
                else:
                    self.logger.info("[AMInterface] How to process " + data)
        except:
            e = sys.exc_info()[0]
            self.logger.info("[AMInterface] error: " + str(e))

    # read string of utf format
    def readUTF(self, sock):
        size = struct.unpack("!H", sock.recv(2))[0]
        msg = sock.recv(size)
        return msg

    # write string as utf format
    def writeUTF(self, sock, msg):
        size = len(msg)
        sock.send(struct.pack("!H", size))
        sock.send(msg)

if __name__ == "__main__":
     AMInterface()

File: agents/apps/onos/onos-1.13.1/src/main/java/org/deltaproject/onosagent/AMInterface.java
package org.deltaproject.onosagent;

import org.slf4j.Logger;

import java.io.BufferedReader;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.EOFException;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.ConnectException;
import java.net.Socket;

import static org.slf4j.LoggerFactory.getLogger;

/**
 * An Agent Manager Interface of ONOS AppAgent.
 */
public class AMInterface extends Thread {

    private AppAgent app;
    private Socket socket;
    private InputStream in;
    private DataInputStream dis;
    private OutputStream out;
    private DataOutputStream dos;
    private String serverIP;
    private int serverPort;
    private final Logger log = getLogger(getClass());

    public AMInterface(AppAgent in) {
        this.app = in;
    }

    public void setServerAddr() {
        // default
        this.serverIP = "10.0.2.2";
        this.serverPort = 3366;

        String path = "~";

        String home = System.getenv("HOME");
        if (home != null) {
            path = home;
        }

        BufferedReader br = null;
        InputStreamReader isr = null;
        FileInputStream fis = null;
        File file = new File(path + "/agent.cfg");
        String temp;

        try {
            fis = new FileInputStream(file);
            isr = new InputStreamReader(fis, "UTF-8");
            br = new BufferedReader(isr);

            while ((temp = br.readLine()) != null) {
                if (temp.contains("MANAGER_IP")) {
                    this.serverIP = temp.substring(temp.indexOf("=") + 1);
                }
                if (temp.contains("MANAGER_PORT")) {
                    this.serverPort = Integer.parseInt(temp.substring(temp.indexOf("=") + 1));
                }
            }

        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            try {
                fis.close();
                isr.close();
                br.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public void connectServer(String agent) throws Exception {
        socket = new Socket(serverIP, serverPort);
        in = socket.getInputStream();
        dis = new DataInputStream(in);
        out = socket.getOutputStream();
        dos = new DataOutputStream(out);

        dos.writeUTF(agent);
        dos.flush();
    }

    public void write(String input) {
        try {
            dos.writeUTF(input);
            dos.flush();
        } catch (IOException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }

    public void replayingKnownAttack(String recv) throws IOException {
        String result = "";

        if (recv.contains("3.1.020")) {
            app.setControlMessageDrop();
            result = app.testControlMessageDrop();
            dos.writeUTF(result);
        } else if (recv.contains("3.1.030")) {
            app.setInfiniteLoop();
            return;
        } else if (recv.contains("3.1.040")) {
            result = app.testInternalStorageAbuse();
            dos.writeUTF(result);
        } else if (recv.contains("3.1.070")) {
            result = app.testFlowRuleModification();
            dos.writeUTF(result);
        } else if (recv.contains("3.1.080")) {
            app.testFlowTableClearance();
            dos.writeUTF(result);
        } else if (recv.contains("3.1.090")) {
            if (app.testEventListenerUnsubscription()) {
                dos.writeUTF("success");
            } else {
                dos.writeUTF("fail");
            }
        } else if (recv.contains("3.1.100")) {
            result = app.testApplicationEviction("fwd");
            dos.writeUTF(result);
        } else if (recv.contains("3.1.110")) {
            app.testResourceExhaustionMem();
            return;
        } else if (recv.contains("3.1.120")) {
            app.testResourceExhaustionCpu();
            return;
        } else if (recv.contains("3.1.130")) {
            app.testSystemVariableManipulation();
            return;
        } else if (recv.contains("3.1.140")) {
            app.testSystemCommandExecution();
            return;
        } else if (recv.contains("3.1.190")) {
            app.testFlowRuleFlooding();
            return;
        } else if (recv.contains("3.1.200")) {
            result = app.testSwitchFirmwareMisuse();
            dos.writeUTF(result);
        } else if (recv.contains("2.1.060")) {

            String cmd = null;
            if (recv.contains("install")) {
                result = app.sendUnFlaggedFlowRemoveMsg("install", 0);
            } else if (recv.contains("check")) {
                long ruleId = Long.parseLong(recv.split("\\|")[2]);
                result = app.sendUnFlaggedFlowRemoveMsg("check", ruleId);
            }
            dos.writeUTF(result);
	} else if (recv.contains("3.1.240")) {
	    result = app.testInfiniteFlowRuleSynchronization();
	    dos.writeUTF(result);
        } else if (recv.contains("test")) {

            return;
        }

        dos.flush();
    }

    public void findingUnkwonAttack(String recv) {

    }

    @Override
    public void run() {
        String recv = "";

        while (true) {
            try {
                this.setServerAddr();
                this.connectServer("AppAgent");
                while (true) {
                    recv = dis.readUTF();
                    log.info("[App-Agent] Received " + recv);
                    if (recv.contains("umode")) {
                        findingUnkwonAttack(recv);
                    } else if (recv.contains("echo")) {
                        write("echo");
                    } else {
                        replayingKnownAttack(recv);
                    }
                }
            } catch (ConnectException e) {
                log.error("[App-Agent] Agent Manager is not listening");
            } catch (EOFException e) {
                // if any error occurs
                log.info("[App-Agent] Closing...");
            } catch (Exception e) {
                log.error(e.toString());
            } finally {
                try {
                    if (dis != null) {
                        dis.close();
                    }
                    if (dos != null) {
                        dos.close();
                    }
                } catch (IOException e) {
                    log.error(e.toString());
                }
            }
            try {
                Thread.sleep(5000);
            } catch (InterruptedException e) {
                log.error(e.toString());
            }
        }
    }
}
